sonar-plugin-server/src/main/java/jetbrains/buildserver/sonarplugin/buildfeatures/BranchesAndPullRequestsParametersPreprocessor.java

```diff
@@ -0,0 +1,107 @@
+            return;
+        }
+
+        if (!isTeamCityMinimalVersion(buildParams.get("system.teamcity.version"))) {
+            build.getBuildLog().message(
+                    String.format("Build feature '%s' requiers TeamCity 2019.2 or above", BranchesAndPullRequestsBuildFeature.BUILD_FEATURE_NAME),
+                    Status.ERROR, MessageAttrs.attrs());
+            return;
+        }
+
+        String branchIsDefault = buildParams.get("teamcity.build.branch.is_default");
+        if (StringUtils.isEmpty(branchIsDefault) || Boolean.TRUE.equals(Boolean.valueOf(branchIsDefault))) {
+            // No information or default branch, nothing to provide
+            return;
+        }
+
+        final String type;
+        final JsonObject json = new JsonObject();
+        final String prNumber = buildParams.get("teamcity.pullRequest.number");
+        if (StringUtils.isEmpty(prNumber)) {
+            // Branch
+            type = "branch";
+            final String vcsBranch = buildParams.get("vcsroot.branch");
+            json.addProperty("sonar.branch.name", buildParams.get("teamcity.build.branch"));
+            json.addProperty("sonar.branch.target", vcsBranch.substring(vcsBranch.indexOf("refs/heads/") + 11));
+        } else {
+            // Pull Request
+            type = "pull-request";
+            String repo = buildParams.get("vcsroot.url");
+            repo = repo.replaceFirst("^git@.*:", "");
+            repo = repo.replaceFirst("^https?://[^/]*/", "");
+            repo = repo.replaceFirst("\\.git$", "");
+
+            json.addProperty("sonar.pullrequest.key", prNumber);
+            json.addProperty("sonar.pullrequest.branch", buildParams.get("teamcity.pullRequest.title"));
+            json.addProperty("sonar.pullrequest.base", buildParams.get("teamcity.pullRequest.target.branch"));
+            json.addProperty("sonar.pullrequest.provider", "github");
+            json.addProperty("sonar.pullrequest.github.repository", repo);
+        }
+
+        final String jsonString = json.toString();
+        build.getBuildLog().message(String.format("SonarQube plugin detects %s, '%s' set with '%s'", type, SQS_SYSENV, jsonString), Status.NORMAL,
+                MessageAttrs.attrs());
+        buildParams.put(SQS_SYSENV, jsonString);
+    }
+
+    /**
+     * Check TeamCity minimal version
+     * 
+     * @param version Current version
+     * @return true if version is 2019.2 or above
+     */
+    static boolean isTeamCityMinimalVersion(String version) {
+        if (StringUtils.isEmpty(version)) {
+            // If absent, version considered as OK
+            return true;
+        }
+
+        String[] v = version.replaceFirst(" ?\\(build.*\\)", "").split("\\.");
+        int major = v.length > 0 ? Integer.valueOf(v[0]) : 0;
+        int minor = v.length > 1 ? Integer.valueOf(v[1]) : 0;
+
+        return major > 2019 || (major == 2019 && minor >= 2);
+    }
```

Comment:
Perhaps it exist a more convenient way to detect TeamCity minimal version ...

Reply:
ServerVersionHolder could help

Reply:
> ServerVersionHolder could help

Fixed by commit [Improve minimal TeamCity version check](https://github.com/JetBrains/TeamCity.SonarQubePlugin/pull/77/commits/57faa0b22946313be91d7233e5e9a91ab870ef84)